tcmur_device.c

```diff
@@ -74,13 +61,28 @@ int __tcmu_reopen_dev(struct tcmu_device *dev, int retries)
 		 */
 		tcmu_dev_flush_ring(dev);
 
+	/*
+	 * To make sure all the in-flight IOs have been finished before
+	 * flushing the event_work. Or just after we flush the event_work
+	 * a new timedout IO callback could come and then it will fire a
+	 * new event work, which may access the device while we are closing
+	 * the device later.
+	 */
 	tcmu_dev_dbg(dev, "Waiting for outstanding commands to complete\n");
 	ret = aio_wait_for_empty_queue(rdev);
```

Comment:
There are two ways to do that: 
1, `TCMUR_DEV_FLAG_IN_RECOVERY`: then all the cmds in the ring buffer will be cleaned by just returning `-EBUSY` without not doing the real IOs to handlers.

So just after `aio_wait_for_empty_queue()`, we are safe to say that all the in-flight IOs have finished.

2, There is one interface to tell the kernel LIO/TCMU do not queue new cmds to the ring buffer and then in userland we can just clean the ring buffer. And then call `aio_wait_for_empty_queue()`.


Here we will use the first one.

Reply:
> So just after `aio_wait_for_empty_queue()`, we are safe to say that all the in-flight IOs have finished.

I did some grepping to confirm that this is the case.  It appears to be but I have two suggestions:
- set `TCMUR_DEV_FLAG_IN_RECOVERY` in `tcmu_add_dev_to_recovery_list` where the device is added to recovery list instead of conditionally in `__tcmu_notify_conn_lost`:
```
add_to_list:
	list_add(&tpg->devs, &rdev->recovery_entry);
	rdev->flags |= TCMUR_DEV_FLAG_IN_RECOVERY;
```
- replace "TCMUR_DEV_FLAG_IN_RECOVERY must be set before calling" comment with an assert as this is critical requirement -- crashing on a NULL dereference somewhere in librados is much worse:
```
	bool needs_close = false;

	pthread_mutex_lock(&rdev->rdev_lock);
	assert(rdev->flags & TCMUR_DEV_FLAG_IN_RECOVERY);
	if (rdev->flags & TCMUR_DEV_FLAG_STOPPING) {
		ret = 0;
		goto done;
```
What do you think?

Reply:
> > So just after `aio_wait_for_empty_queue()`, we are safe to say that all the in-flight IOs have finished.
> 
> I did some grepping to confirm that this is the case. It appears to be but I have two suggestions:
> 
> * set `TCMUR_DEV_FLAG_IN_RECOVERY` in `tcmu_add_dev_to_recovery_list` where the device is added to the recovery list instead of conditionally in `__tcmu_notify_conn_lost`:
> 
> ```
> add_to_list:
> 	list_add(&tpg->devs, &rdev->recovery_entry);
> 	rdev->flags |= TCMUR_DEV_FLAG_IN_RECOVERY;
> ```
> 

Yeah, this looks okay for me.


> * replace "TCMUR_DEV_FLAG_IN_RECOVERY must be set before calling" comment with an assert as this is critical requirement -- crashing on a NULL dereference somewhere in librados is much worse:
> 
> ```
> 	bool needs_close = false;
> 
> 	pthread_mutex_lock(&rdev->rdev_lock);
> 	assert(rdev->flags & TCMUR_DEV_FLAG_IN_RECOVERY);
> 	if (rdev->flags & TCMUR_DEV_FLAG_STOPPING) {
> 		ret = 0;
> 		goto done;
> ```
> 
> What do you think?

This also makes sense to me.

I will revise this. thanks.